### ISSUE TITLE
Relations in a Non Relational database

### ISSUE THREAD

**Issue by angyts** [susanBuck/e28-spring20#64]
Dear Wonderful Classmates,

I know this class is totally not about this. But I have a small question because it is related to my project 2. I'm doing some kind of scheduling and calendar, it it has proven quite daunting... I am on the brink of giving it all up and doing something easier...

Anyway, I'm using firebase as a backend, which is something new for me.

I have these Collection: `shifts` which is like a time slot, which I want to schedule people to work inside them, like `employees`.

Normally in relational database, it would be like `shifts.employee_working` --> foreign key ---> `employees`.

Now I have to store shifts like this:
```
one_instance_of_a_shift: {
    'date': ...,
    'time':...,
    'employees': [
        'personA' : { name:...,
                           id: ...,
                          pic: ...,
      },
        'personB': { name:...,
                           id: ...,
                          pic: ...,
      },
 ]
}
```

Wouldn't I keep replicating the employees all over the database? Is this how it is supposed to work?

**Comment by DoryAzar** [comment]
It is true that you would be replicating the employees and I believe in real-life this is most likely not desirable but for the purpose of this project, I believe it is acceptable. Generally speaking, in noDB models, there is an equivalent for creating relationships between collections which usually is done by references. I believe Firestore has that as well: 
https://firebase.google.com/docs/firestore/data-model

You probably wanna check that direction if you want to get into it.

**Comment by susanBuck** [comment]
Adding to @DoryAzar's notes - I'd say that "NoSQL" solutions like Firestore is probably not the most appropriate solution for the kind of relationship-heavy data structure you're describing. 

To this point, I'd note that Firestore is just one possible set up you can use for your server API/data source (and the one we happen to be exploring), but it's not always the best tool for the job. Depends a lot on what you're building, scale, concerns about size vs. performance, etc.

To get something set up for the sake of the project and what you're aiming for, I'd definitely check out the Firestore **references** @DoryAzar pointed to.

**Comment by angyts** [comment]
I actually did try the "relations".

1) Firstly, the firebase documentation is a little skimpy. (I came from Django which has IMO the best documentation in techworld, followed by Vue)
2) It seems like I still have to retrieve the "relations" myself. Meaning that after I retrieve the main document, I still have to retrieve the "related" document via another call. Which is really a waste of time., i might as well just store the ID of the related document as a string. 
3) So I think yes, storing extra information seems like the way to go, and perhaps at least keep the foreign key document ID constant. So if the foreign key `user` is actually updated, I can probably trigger some `post-save` method to update all the relations table with the correct details.
